### PR TITLE
Feature/tracer

### PR DESCRIPTION
**Pull request by fpdotmonkey**
Made an interface over which computed data can go to a web-based front end.

### CODE REVIEW

File: Display/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+RUN go get github.com/pebbe/zmq4 && go get github.com/gorilla/websocket
+
+RUN mkdir -p /var/log/supervisor
+RUN mkdir -p /src/zmq_proxy/static
+WORKDIR /src/zmq_proxy
+COPY main.go /src/zmq_proxy
+COPY static /src/zmq_proxy/static
+RUN go build
+
+COPY requirements.txt /src/zmq_proxy
```
Comment (mmroden) [inline Display/Dockerfile:18]: If you go the route of copying individual files vs a whole directory, you have to make sure to copy new files when they get made, or else the docker image won't contain your changes.  As such, you should run tests against the docker image, which is the final deployable artifact, rather than the code itself.  (You may already be doing this, but I'm just starting the review).

File: Display/features/calculating_tubing_descriptors.feature [hunk 2]
```diff
@@ -0,0 +1,11 @@
+Feature: Calculating tubing descriptors
+
+    Scenario: Calculating descriptors
+        Given we can use sensor data
```
Comment (mmroden) [inline Display/features/calculating_tubing_descriptors.feature]: These kinds of tests are generally something like "Given we have this set of objects, when this data is given to the objects, then these results happen".  You've combined steps here.  As the test suite expands, you may want the steps to be separated so that you can reuse steps.  For instance, if you are always instantiating the same set of objects and storing them in the context, then that set can be in the givens.  _When_ you calculate the descriptors, _then_ the results of a particular set of input data match your expectations.  Right now, this reads that you're calculating things, but you're not checking the output of the calculations for correctness.

File: Display/features/gets_sensor_data.feature [hunk 3]
```diff
@@ -0,0 +1,15 @@
+Feature: gets data from sensors
+
+    Scenario: connects to the sensors that are connected
+        Given some sensors are physically connected to the Pressure Sensor and Airflow Sensor PI Hat
```
Comment (mmroden) [inline Display/features/gets_sensor_data.feature]: I would word this test as 

- Given there are these objects that represent these sensors
- When data is read from those sensors on startup
- then the software will perform these actions when all is well

That allows you to create mock interfaces to a dummy sensor object that can either spit out the "right" values or garbage, depending on what you want to test.

File: Display/features/gets_sensor_data.feature [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+        Then the software will determine which tubes have both pressure and airflow sensing
+        Then the software will initialize and calibrate the sensors
+        Then the software will return the state of the sensor to the user.
+
+    Scenario: there are not enough sensors
+        Given there are no tubes with both pressure and airflow sensing connected
```
Comment (mmroden) [inline Display/features/gets_sensor_data.feature]: - Given a set of functioning sensors
- and a set of malfunctioning sensors or parts
- when the software diagnostic is run
- then this error will be raised

With the above approach, you can actually have a pretty complete set of tests to say things like "Flow sensor 2 appears to be malfunctioning", nomsain?

File: Display/features/gets_sensor_data.feature [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+        Then the software will raise an error
+
+    Scenario: there are multiple airflow sensors on a single tube
+        Given more than one airflow sensor is connected to a single tube
```
Comment (mmroden) [inline Display/features/gets_sensor_data.feature]: - Given sensor A on a tube
- and less functional sensor B on a tube
- When data is read
- then the data from A will be used

Something like this means you can have lots of different sensor "givens" and reuse them, and have a "when" that is "reading the data" that you reuse between tests that goes through all the initialized sensors in the context.

You will need to define what "worse" is here,

File: Display/features/steps/calculating_tubing_descriptors.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+from behave import *
+
+import server
+
+@given("we can use sensor data")
+def step_impl(context):
+
```
Comment (mmroden) [inline Display/features/steps/calculating_tubing_descriptors.py:15]: Is this space in pep8?  Because there's spaces in pep8, and I don't think this space is there, and there's two spaces between functions.

File: Display/server.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
+import sys
+import time
+import json
+import random
+
+import zmq
+
+from patienttubingdescriptorcalculator \
+    import PatientTubingDescriptorCalculator
```
Comment (mmroden) [inline Display/server.py:9]: you don't _have_ to comply to the 80 character line limit now that we have widescreen monitors with more than 1024x768 resolution.

File: Display/server.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        self._socket.bind(f"tcp://*:{port}")
+
+    def publish_message(self, message):
+        
+        self._socket.send_multipart([b"",
+                                     json.dumps(message).encode("ascii")])
+
+def main():
+
+    NUMBER_OF_PATIENTS = 4
```
Comment (mmroden) [inline Display/server.py]: constants should go to the top of the file, or better yet, into a single constants file you import

**Review by mmroden — COMMENTED** [review]
Just some general commentary on the behave tests

File: Display/server.py (hunk 7 above)
Comment (fpdotmonkey) [reply Display/server.py:9]: Ah, but PEP8 says _79_ columns
"Limit all lines to a maximum of 79 characters."

File: Display/features/steps/calculating_tubing_descriptors.py (hunk 6 above)
Comment (fpdotmonkey) [reply Display/features/steps/calculating_tubing_descriptors.py:15]: I added the blank based on "Method definitions inside a class are surrounded by a single blank line."  Granted this isn't in a class.  Does "method definition" in this context mean the `def` line, or the whole function block?
Actually looking at PEP8 again, looks like I was supposed to have 2 blank lines: "Surround top-level function and class definitions with two blank lines."

File: Display/Dockerfile (hunk 1 above)
Comment (fpdotmonkey) [reply Display/Dockerfile:18]: To confirm I'm understanding right, you're saying I should run the tests from within the docker image, yeah?

Comment (mmroden) [reply Display/Dockerfile:18]: no, I'm saying that when you start your behave tests, you should set up the environment to run the docker image, and query the docker image from the outside.  If you don't, then you run the risk of having a file outside of the docker that's not in the docker, and the tests passing, but the docker doesn't.

File: Display/features/steps/calculating_tubing_descriptors.py (hunk 6 above)
Comment (mmroden) [reply Display/features/steps/calculating_tubing_descriptors.py:15]: yeah, these aren't inside classes, but the 'def' marks the top.  Even so, either way, there isn't a space between the definition line and the contents of the function, but you can put comments in there if you'd like.

File: Display/server.py (hunk 7 above)
Comment (mmroden) [reply Display/server.py:9]: and I'm saying that line can be ignored because we have wider screens than 1024x768.

File: Display/features/environment.py [hunk 9]
```diff
@@ -0,0 +1,8 @@
+
+
```
Comment (mmroden) [inline Display/features/environment.py:2]: This is where you can do 'before_all' and start your docker container, to do testing external to the code

File: Display/features/steps/runs_backend_from_docker_image.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
+from behave import given, when, then
```
Comment (mmroden) [inline Display/features/steps/runs_backend_from_docker_image.py:1]: thank you :)

File: Display/main.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    subscribe string
+}
+
+type Client struct {
+    conn *websocket.Conn
+}
+
+func (client *Client) readPump() {
+    defer func() {
+        log.Printf("read pump closing")
```
Comment (mmroden) [inline Display/main.go:32]: Logging is a big thing in python: https://docs.python.org/3/library/logging.html

If you're going to be doing logging, make that a separate PR, I think.

File: Display/sensors.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    @abstractmethod
+    def calibration_pressure_sensor_connected(self):
+        pass
+
+    @abstractmethod
+    def poll(self):
+        pass
+
+
+if is_on_raspberry_pi():
```
Comment (mmroden) [inline Display/sensors.py:33]: clever use of an if; does this work?

File: Display/features/steps/gets_sensor_data.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+@given("there is an object that represent the total sensing package")
+def step_impl(context):
+    os.environ[constants.SENSOR_QUANTITY] = constants.ENOUGH_SENSORS
+    context.sensors = server.Sensors(1, 1, 1, 1)
+
+
+@when("these sensors are initialized")
+def step_impl(context):
+    pass
```
Comment (mmroden) [inline Display/features/steps/gets_sensor_data.py:17]: are sensors automatically initialilzed?

File: Display/features/steps/gets_sensor_data.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+def step_impl(context, sensor):
+    pass
+
+
+@when("data is read")
+def step_impl(context):
+    pass
+
+
+@then("the data from {sensor} will be used")
```
Comment (mmroden) [inline Display/features/steps/gets_sensor_data.py:74]: This step would be "the data matches my expectation that looks like XYZ"

File: Display/sensors.py (hunk 12 above)
Comment (fpdotmonkey) [reply Display/sensors.py:33]: It does, yeah.  I copied some code from the RPi standard library (with attribution) to ID the platform.  It can even negatively identify similar boards like Beaglebone Blacks.

File: Display/features/steps/gets_sensor_data.py (hunk 13 above)
Comment (fpdotmonkey) [reply Display/features/steps/gets_sensor_data.py:17]: My thinking was that the drivers would initialize would initialize the sensors and that that step would be hidden from the person using the `sensors` module.  But I'm not totally sure.  This is one of those instances where I got stuck thinking I need to write the test and define the code, but I need an idea of what the code is going to be in order to write the test.
